pkg/gadgets/seccomp/gadget.go

```diff
@@ -0,0 +1,263 @@
+	if trace.ObjectMeta.Namespace != gadgets.TRACE_DEFAULT_NAMESPACE {
+		trace.Status.OperationError = fmt.Sprintf("This gadget only accepts operations on traces in the %s namespace", gadgets.TRACE_DEFAULT_NAMESPACE)
+		return
+	}
+	if trace.ObjectMeta.Name != "seccomp" {
+		trace.Status.OperationError = fmt.Sprintf("This gadget only accepts operations on traces named %s", "seccomp")
+		return
+	}
+	switch operation {
+	case "start":
+		t.Start(trace)
+	case "generate":
+		t.Generate(trace, resolver, params)
+	case "stop":
+		t.Stop(trace)
+	default:
+		trace.Status.OperationError = fmt.Sprintf("Unknown operation %q", operation)
+	}
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		trace.Status.OperationError = ""
+		trace.Status.Output = ""
+		trace.Status.State = "Started"
+		return
+	}
+
+	t.cmd = exec.Command("/bin/seccomp-policy-advisor")
+	t.out.Reset()
+	t.cmd.Stdout = &t.out
+	err := t.cmd.Start()
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to start: %s", err)
+		return
+	}
+	t.started = true
+
+	trace.Status.OperationError = ""
+	trace.Status.Output = ""
+	trace.Status.State = "Started"
+	return
+}
+
+func (t *Trace) Generate(trace *gadgetv1alpha1.Trace, resolver gadgets.Resolver, params map[string]string) {
+	if !t.started {
+		trace.Status.OperationError = "Not started"
+		return
+	}
+
+	pod, ok := params["pod"]
+	if !ok {
+		trace.Status.OperationError = "Error: pod parameter missing"
+		return
+	}
+	mntns, err := resolver.LookupMntnsByNamespacedName(pod)
+	if !ok {
+		trace.Status.OperationError = fmt.Sprintf("Error: pod %q not found: %s", pod, err)
+		return
+	}
+
+	mapSpec := &ebpf.MapSpec{
+		Name:       BPF_MAP_NAME,
+		Type:       ebpf.Hash,
+		KeySize:    8,
+		ValueSize:  C.SYSCALLS_MAP_VALUE_SIZE,
+		MaxEntries: 1024,
+		Pinning:    ebpf.PinByName,
+	}
+	mapInstance, err := ebpf.NewMapWithOptions(mapSpec, ebpf.MapOptions{PinPath: gadgets.PIN_PATH})
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to create map %s: %s", BPF_MAP_NAME, err)
+		return
+	}
+
+	b, err := mapInstance.LookupBytes(mntns)
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to find any syscalls for pod %q: %s", pod, err)
+		return
+	}
+
+	policy := syscallArrToLinuxSeccomp(b)
+
+	output, err := json.MarshalIndent(policy, "", "  ")
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to marshal seccomp policy for pod %q: %s", pod, err)
+		return
+	}
+
+	trace.Status.Output = string(output)
+	trace.Status.OperationError = ""
+}
+
+func (t *Trace) Stop(trace *gadgetv1alpha1.Trace) {
+	if !t.started {
+		trace.Status.OperationError = "Not started"
+		return
+	}
+	err := t.cmd.Process.Signal(syscall.SIGINT)
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to send SIGINT to process: %s", err)
+		return
+	}
+
+	err = t.cmd.Wait()
```

Comment:
What if the process is somehow hung and won't stop with sigint? Shouldn't this have some timeout?

Reply:
Yes... I am not quite sure we're allowed to use one of the controller threads to run blocking operations.

I am currently rewriting this code to implement the seccomp gadget with a Go package instead of using an external binary (suggestion from @mauriciovasquezbernal ). So it will not be an issue after the refactoring. But the same comment applies to other gadgets like biolatency, so maybe we'll need to implement the timeout there.

Reply:
Refactoring done, so the seccomp gadget does not use commands anymore.